README.md
# Magento Cookbook
## Cookbook to provision the server for Magento Framework

The Provision, have many steps to accomplish the tasks.

See Changelog [here](CHANGELOG.md)


CHANGELOG.md
# Changelog

## [fixups](https://github.com/adminstudio/chef-magento/tree/fixups) (2020-04-23)

[Full Changelog](https://github.com/adminstudio/chef-magento/compare/beda4e2806a0f2f1a804e9d039c91ba3b5f2b01b...fixups)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


spec/spec_helper.rb
# frozen_string_literal: true

require 'chefspec'
require 'chefspec/policyfile'

RSpec.configure do |config|
  config.color = true
  config.formatter = :documetation
  config.log_level = :error
end


spec/unit/recipes/default_spec.rb
#
# Cookbook:: magento
# Spec:: default
#
# Copyright:: 2020, Asdrubal Gonzalez Penton, All Rights Reserved.

require 'spec_helper'

describe 'magento::default' do
  context 'When all attributes are default, on Ubuntu 18.04' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'ubuntu', '18.04'

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'centos', '7'

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
    it 'update the system' do
      expect(chef_run).to run_execute('yum update -y')
    end
    it 'install epel repo and wget' do
      expect(chef_run).to install_yum_package('wget')
      expect(chef_run).to install_yum_package('epel-release')
    end
    it 'remi repo present' do
      expect(chef_run).to add_yum_repository('remi')
    end
    it 'remi-safe present' do
      expect(chef_run).to add_yum_repository('remi-safe')
    end
    it 'remi-php73 present' do
      expect(chef_run).to add_yum_repository('remi-php72')
    end
    it 'nginx present' do
      expect(chef_run).to add_yum_repository('nginx-stable')
    end
    it 'update the system' do
      expect(chef_run).to run_execute('yum update -y')
    end
    it 'packages installed' do
      %w[perl unzip net-tools perl-libwww-perl perl-LWP-Protocol-https perl-GDGraph git ccze htop vim glances].each do |pkg|
        expect(chef_run).to install_yum_package(pkg)
      end
    end
    it 'download the virtualmin script installation' do
      expect(chef_run).to create_remote_file('/opt/install.sh')
    end
    it 'install virtualmin' do
      expect(chef_run).to run_execute('install_virtuamin')
    end
    it 'csf downloaded' do
      expect(chef_run).to create_if_missing_remote_file('/opt/csf.tgz')
    end
    it 'csf unzipped' do
      expect(chef_run).to run_execute('unzip_csf')
    end
    it 'csf installed' do
      expect(chef_run).to run_execute('install_csf')
    end
    it 'template applied' do
      expect(chef_run).to render_file('/etc/csf/csf.conf')
    end
    it 'csf running' do
      expect(chef_run).to start_service('csf')
    end
    it 'csf enabled' do
      expect(chef_run).to enable_service('csf')
    end
    it 'install webmin module' do
      expect(chef_run).to run_execute('webmin_module')
    end
    it 'download the netdata script installation' do
      expect(chef_run).to create_if_missing_remote_file('/opt/kickstart.sh')
    end
    it 'install netdata' do
      expect(chef_run).to run_execute('install_netdata')
    end
    it 'files delete from /opt/csf.tgz' do
      expect(chef_run).to delete_file('/opt/csf.tgz')
    end

    it 'files delete from /opt/install.sh' do
      expect(chef_run).to delete_file('/opt/install.sh')
    end

    it 'files delete from /opt/composer.phar' do
      expect(chef_run).to delete_file('/opt/composer.phar')
    end

    it 'files delete from /opt/composer-setup.php' do
      expect(chef_run).to delete_file('/opt/composer-setup.php')
    end
    it 'files delete from /opt/csf' do
      expect(chef_run).to delete_directory('/opt/csf')
    end
  end
end


test/integration/default/default_test.rb
# frozen_string_literal: true

# InSpec test for recipe magento::default
# The InSpec reference, with examples and extensive documentation, can be
# found at https://www.inspec.io/docs/reference/resources/

unless os.windows?
  # This is an example test, replace with your own test.
  describe user('root'), :skip do
    it { should exist }
  end
end

# This is an example test, replace it with your own test.
describe port(80), :skip do
  it { should_not be_listening }
end
describe yum.repo('epel') do
  it { should exist }
  it { should be_enabled }
end

describe yum do
  # its('repos') { should exist }
  its('repos') { should include 'base/7/x86_64' }
  its('repos') { should include 'remi' }
  its('repos') { should include 'remi-safe' }
  # its('repos') { should include 'mariadb' }
  # its('repos') { should include 'virtualmin/7/x86_64' }
end
describe yum.repo('remi') do
  it { should exist }
  it { should be_enabled }
end
describe yum.repo('remi-safe') do
  it { should exist }
  it { should be_enabled }
end
%w[perl unzip net-tools perl-libwww-perl perl-LWP-Protocol-https perl-GDGraph git].each do |pre|
  describe package(pre) do
    it { should be_installed }
  end
end

describe file('/opt/install.sh') do
  it { should_not exist }
  # it { should be_file }
  # it { should be_owned_by 'root' }
  # its('owner') { should eq 'root' }
  # its('mode') { should cmp '0777' }
  # its('type') { should cmp 'file' }
end
describe command('php -v') do
  # its('stdout') { should match 'PHP 7.3./\d{1,2}/' }
  its('stdout') { should match /PHP 7.2.\d{1,2}/ }
end
describe service('webmin') do
  it { should be_installed }
  # it { should be_running }
end

describe file('/opt/csf.tgz') do
  it { should_not exist }
end

describe directory('/opt/csf') do
  it { should_not exist }
end
describe file('/etc/csf/csf.conf') do
  it { should exist }
  # its('stdout') { should include 'RESTRICT_SYSLOG = "3"'}
  # its('stdout') { should include 'SYSLOG_CHECK = "3600"'}
end
describe package('MariaDB-server') do
  it { should be_installed }
end
describe command('mysql --version') do
  its('stdout') { should include '10.3.22-MariaDB' }
end

describe firewalld do
  it { should_not be_running }
end
# describe command('php -i') do
#   its('stdout') { should include 'short_open_tag On'}
# end
describe service('netdata') do
  it { should be_running }
end
describe http('http://localhost:19999') do
  its('status') { should eq 200 }
end


attributes/others.rb
# frozen_string_literal: true

default['virtualminLEMP']['date'] = Time.new
default['virtualminLEMP']['csf']['TESTING'] = '"1"'
default['virtualminLEMP']['csf']['RESTRICT_SYSLOG'] = '"3"'
default['virtualminLEMP']['csf']['LF_POP3D'] = '"10"'
default['virtualminLEMP']['csf']['LF_IMAPD'] = '"10"'
default['virtualminLEMP']['csf']['LF_IMAPD'] = '"10"'
default['virtualminLEMP']['csf']['SYSLOG_CHECK'] = '"3600"'
default['virtualminLEMP']['csf']['TCP_IN'] = '"22,80,443,19999"'
default['virtualminLEMP']['csf']['UDP_IN'] = '"53"'
default['virtualminLEMP']['csf']['TCP6_IN'] = '"80,443,19999"'
default['virtualminLEMP']['csf']['UDP6_IN'] = '"53"'

# my.cnf
default['virtualminLEMP']['mysql']['datadir'] = '/var/lib/mysql'
default['virtualminLEMP']['mysql']['socket'] = '/var/lib/mysql/mysql.sock'
default['virtualminLEMP']['mysql']['log-error'] = '/var/log/mariadb/mariadb_error.log'
default['virtualminLEMP']['mysql']['symbolic-links'] = '0'
default['virtualminLEMP']['mysql']['max_connections'] = '700'
default['virtualminLEMP']['mysql']['join_buffer_size'] = '10M'
default['virtualminLEMP']['mysql']['performance_schema'] = 'on'
default['virtualminLEMP']['mysql']['slow-query-log'] = '1'
default['virtualminLEMP']['mysql']['slow-query-log-file'] = '/var/lib/mysql/mysql-slow.log'
default['virtualminLEMP']['mysql']['long_query_time'] = '1'
default['virtualminLEMP']['mysql']['query_cache_size'] = '128M'
default['virtualminLEMP']['mysql']['query_cache_limit'] = '128M'
default['virtualminLEMP']['mysql']['query_cache_type'] = '1'
default['virtualminLEMP']['mysql']['thread_cache_size'] = '8'
default['virtualminLEMP']['mysql']['myisam_sort_buffer_size'] = '64M'
default['virtualminLEMP']['mysql']['read_rnd_buffer_size'] = '8M'
default['virtualminLEMP']['mysql']['read_buffer_size'] = '2M'
default['virtualminLEMP']['mysql']['sort_buffer_size'] = '2M'
default['virtualminLEMP']['mysql']['table_open_cache'] = '1024'
default['virtualminLEMP']['mysql']['max_allowed_packet'] = '32M'
default['virtualminLEMP']['mysql']['key_buffer_size'] = '384M'
default['virtualminLEMP']['mysql']['tmp_table_size'] = '2G'
default['virtualminLEMP']['mysql']['max_heap_table_size'] = '512M'
default['virtualminLEMP']['mysql']['innodb_file_per_table'] = '2'
default['virtualminLEMP']['mysql']['innodb_log_file_size'] = '2G'
default['virtualminLEMP']['mysql']['innodb_buffer_pool_instances'] = '1'
default['virtualminLEMP']['mysql']['innodb-ft-min-token-size'] = '2'
default['virtualminLEMP']['mysql']['innodb_ft_enable_stopword'] = '""'
default['virtualminLEMP']['mysql']['ft_min_word_len'] = '1'
default['virtualminLEMP']['mysql']['ft_stopword_file'] = '""'
# default['virtualminLEMP']['mysql']['innodb_buffer_pool_size'] = ((1024 + (256 * Math.log(19, 2)) * 2) ).
p_mem = node['memory']['total']
men = p_mem[0...-2].to_i / 1024
t_mem = men
default['virtualminLEMP']['mysql']['innodb_buffer_pool_size'] = ((t_mem * 1024 - 2 * (256 + 256 * Math.log2(t_mem)).to_i) / (1.05 * 1024)).to_i
# if(ramValue <= 1) { result = 128; display = result + ' MB'; } else if(ramValue < 32) { var OS= Math.round(256+256*(Math.log2(ramValue))); result = (ramValue*1024-2*OS)/(1.05*1024); if(result < 2) { display = result.toFixed(2) + ' GB'; } else { display = Math.round(result) + ' GB'; } } else { var OS = Math.round(0.05*ramValue*1024); result = (ramValue*1024-2*OS)/(1.05*1024); display = Math.round(result) + ' GB'; }

# php.ini
default['virtualminLEMP']['php']['short_open_tag'] = 'On'
default['virtualminLEMP']['php']['memory_limit'] = '1024M'
default['virtualminLEMP']['php']['post_max_size'] = '128M'
default['virtualminLEMP']['php']['upload_max_filesize'] = '64M'
# default['virtualminLEMP']['php']['disable_functions'] = 'exec,passthru,shell_exec,system,proc_open,popen,curl_exec,curl_multi_exec,parse_ini_file,show_source'
default['virtualminLEMP']['php']['disable_functions'] = '""'

# nginx.conf

default['virtualmin']['nginx']['user'] = 'nginx'
default['virtualmin']['nginx']['worker_processes'] = 'auto'
default['virtualmin']['nginx']['error_log'] = '/var/log/nginx/error.log'
default['virtualmin']['nginx']['error_log_type'] = 'warn'
default['virtualmin']['nginx']['worker_connections'] = '2048'
default['virtualmin']['nginx']['access_log'] = '/var/log/nginx/access.log'
default['virtualmin']['nginx']['sendfile'] = 'on'
default['virtualmin']['nginx']['log_subrequest'] = 'on'
default['virtualmin']['nginx']['tcp_nopush'] = 'on'
default['virtualmin']['nginx']['tcp_nodelay'] = 'on'
default['virtualmin']['nginx']['etag'] = 'on'
default['virtualmin']['nginx']['keepalive_timeout'] = '75'
default['virtualmin']['nginx']['types_hash_max_size'] = '2048'
default['virtualmin']['nginx']['resolver'] = '1.1.1.1 valid=30s'
default['virtualmin']['nginx']['gzip'] = 'on'
default['virtualmin']['nginx']['gzip_comp_level'] = '6'
default['virtualmin']['nginx']['gzip_min_length'] = '100'
default['virtualmin']['nginx']['gzip_http_version'] = '1.1'
default['virtualmin']['nginx']['gzip_vary'] = 'on'
default['virtualmin']['nginx']['gzip_proxied'] = 'any'
default['virtualmin']['nginx']['gzip_types'] = 'text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript image/svg+xml application/x-font-ttf font/opentype image/jpeg image/png'
default['virtualmin']['nginx']['gzip_static'] = 'always'
default['virtualmin']['nginx']['gzip_buffers'] = '16 8k'
default['virtualmin']['nginx']['gzip_disable'] = '“MSIE [1-6].(?!.*SV1)”'
default['virtualmin']['nginx']['server_tokens'] = 'off'
default['virtualmin']['nginx']['fastcgi_read_timeout'] = '600'
default['virtualmin']['nginx']['server_names_hash_bucket_size'] = '128'
default['virtualmin']['nginx']['cloudflare_ips'] = '/etc/nginx/cloudflare.conf'


recipes/default.rb
# frozen_string_literal: true

# Cookbook:: magento
# Recipe:: default
#
# Copyright:: 2020, Asdrubal Gonzalez Penton, All Rights Reserved.

# host_name = node['fqdn']

# Update the system

execute 'yum update -y' do
  action :run
  ignore_failure true
end

# Install EPEL Repo and wget for get some scripts
%w[epel-release wget].each do |pkg|
  yum_package(pkg) do
    action :install
  end
end

# get the gpgkey to the node.
remote_file '/etc/pki/rpm-gpg/RPM-GPG-KEY-remi' do
  source 'https://rpms.remirepo.net/RPM-GPG-KEY-remi'
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

yum_repository 'remi' do
  description                "Remi's RPM repository for Enterprise Linux 7 - $basearch"
  enabled                    true
  fastestmirror_enabled      true
  gpgcheck                   true
  gpgkey                     'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-remi'
  mirrorlist                 'http://cdn.remirepo.net/enterprise/7/remi/mirror'
  action                     :add
end

yum_repository 'remi-safe' do
  description                "Remi's RPM repository for Enterprise Linux 7 - $basearch"
  enabled                    true
  fastestmirror_enabled      true
  gpgcheck                   true
  gpgkey                     'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-remi'
  mirrorlist                 'http://cdn.remirepo.net/enterprise/7/remi/mirror'
  action                     :add
end

yum_repository 'remi-php72' do
  description                "Remi's PHP 7.2 RPM repository for Enterprise Linux 7 - $basearch"
  enabled                    true
  fastestmirror_enabled      true
  gpgcheck                   true
  gpgkey                     'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-remi'
  mirrorlist                 'http://cdn.remirepo.net/enterprise/7/php72/mirror'
  action                     :add
end
# remote_file '/opt/rabbitmq-release-signing-key.asc' do
#   source 'https://github.com/rabbitmq/signing-keys/releases/download/2.0/rabbitmq-release-signing-key.asc'
#   owner 'root'
#   group 'root'
#   mode '0755'
#   action :create
#   # notifies :install, 'execute[rabbitmq_key]', :immediately
# end
#
# execute 'rpm --import /opt/rabbitmq-release-signing-key.asc'

yum_repository 'bintray-rabbitmq-server' do
  description 'bintray-rabbitmq-rpm'
  enabled true
  # gpgkey 'https://github.com/rabbitmq/signing-keys/releases/download/2.0/rabbitmq-release-signing-key.asc'
  gpgcheck false
  baseurl 'https://dl.bintray.com/rabbitmq/rpm/rabbitmq-server/v3.8.x/el/7/'
end

remote_file '/opt/mysql80-community-release-el7-3.noarch.rpm' do
  source 'https://dev.mysql.com/get/mysql80-community-release-el7-3.noarch.rpm'
  owner 'root'
  group 'root'
  mode '0755'
  action :create
  notifies :install, 'rpm_package[mysql_community]', :immediately
end
rpm_package 'mysql_community' do
  package_name 'mysql80-community-release-el7-3.noarch.rpm'
  source '/opt/mysql80-community-release-el7-3.noarch.rpm'
  action :nothing
  notifies :install, 'yum_package[mysql-community-server]', :immediately
end

yum_package 'mysql-community-server' do
  action :nothing
end
# rpm --import https://github.com/rabbitmq/signing-keys/releases/download/2.0/rabbitmq-release-signing-key.asc
# [bintray-rabbitmq-server]
# name=bintray-rabbitmq-rpm
# baseurl=https://dl.bintray.com/rabbitmq/rpm/rabbitmq-server/v3.8.x/el/7/
# gpgcheck=0
# repo_gpgcheck=0
# enabled=1

# yum_repository 'mariadb' do
#   description 'MariaDB 10.3 CentOS repository'
#   enabled true
#   gpgkey 'https://yum.mariadb.org/RPM-GPG-KEY-MariaDB'
#   gpgcheck true
#   baseurl 'http://yum.mariadb.org/10.3/centos7-amd64'
# end

# mysql 5.7 https://downloads.mysql.com/archives/get/p/23/file/mysql-community-server-5.7.28-1.el7.x86_64.rpm
# mysql 8 https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-community-server-8.0.19-1.el7.x86_64.rpm
# Signature Checking Using RPM
# rpm --checksig package_name.rpm
# mysql repo https://dev.mysql.com/get/mysql80-community-release-el7-3.noarch.rpm
# mysql apt repo https://dev.mysql.com/get/mysql-apt-config_0.8.15-1_all.deb

yum_repository 'nginx-stable' do
  description                'nginx stable repo'
  enabled                    true
  fastestmirror_enabled      true
  gpgcheck                   true
  gpgkey                     'https://nginx.org/keys/nginx_signing.key'
  baseurl                    'http://nginx.org/packages/centos/$releasever/$basearch/'
  action                     :add
  # # TODO: module_hotfixes=true
  notifies :run, 'execute[yum update -y]', :immediately
  notifies :install, 'yum_package[nginx]', :immediately
end

yum_package 'nginx' do
  action :install
end

# %w[unzip net-tools git gzip lsof mysqldump nice sed tar].each do |deps|
%w[unzip net-tools git gzip lsof zip libtool mailx ccze htop vim glances].each do |deps|
  yum_package deps do
    action :install
    # notifies :notify, 'slack_notify[end_preparation]', :immediately
  end
end

%w[perl perl-libwww-perl perl-LWP-Protocol-https perl-GDGraph].each do |perlp|
  yum_package perlp do
    action :install
    # notifies :notify, 'slack_notify[end_preparation]', :immediately
  end
end

# slack_notify 'end_preparation' do
#   message "The preparation is done, let's setup the Server #{host_name}"
#   webhook_url 'https://hooks.slack.com/services/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'
#   action :nothing
# end

group 'magento' do
  action :create
  # members 'magento'
  # append true
end

user 'magento' do
  comment 'the user for magento site'
  uid 1001
  gid 'magento'
  home '/home/magento'
  manage_home true
  shell '/bin/bash'
  action :create
  # password '$1$JJsvHslasdfjVEroftprNn4JHtDi'
end
group 'nginx' do
  action :modify
  members 'magento'
  append true
end

%w[php-mysql php-gd php-fpm php-pear php-bcmath php-mcrypt php-intl php-ctype php-curl php-dom php-hash php-iconv php-mbstring php-openssl php-pdo_mysql php-simplexml php-soap php-xsl php-zip php-libxml].each do |phpmod|
  yum_package phpmod do
    action :install
  end
end

# %w[httpd-devel libtool mailx redis ccze htop vim glances].each do |extra|
#   yum_package extra do
#     action :install
#   end
# end

template '/etc/php.ini' do
  source 'php.ini.erb'
  mode '0644'
  owner 'root'
  group 'root'
  variables(
    open_tag: node['magento']['php']['short_open_tag'],
    disable_functions: node['magento']['php']['disable_functions'],
    memory_limit: node['magento']['php']['memory_limit'],
    post_max_size: node['magento']['php']['post_max_size'],
    upload_max_filesize: node['magento']['php']['upload_max_filesize']
  )
end

service 'php-fpm' do
  subscribes :restart, 'file[/etc/php.ini]', :immediately
  # notifies :notify, 'slack_notify[php_configured]', :immediately
end

remote_file '/opt/composer-setup.php' do
  source 'https://getcomposer.org/installer'
  owner 'root'
  group 'root'
  action :create_if_missing
end

execute 'create_installer' do
  cwd '/opt'
  command 'php /opt/composer-setup.php --quiet'
end

remote_file 'install_composer' do
  path '/usr/local/bin/composer'
  source 'file:///opt/composer.phar'
  owner 'root'
  group 'root'
  mode '0777'
end

execute 'mkdir /opt/scripts' do
  action :run
  not_if { Dir.exist?('/opt/scripts') }
end

cookbook_file '/opt/scripts/cloudflare_ips.sh' do
  source 'cloudflare_ips.sh'
  owner 'root'
  group 'root'
  mode '0777'
  action :create
  # notifies :run, 'execute[add_cloudflare_ips]', :immediately
end

# execute 'add_cloudflare_ips' do
#   cwd '/opt/scripts'
#   command 'sh cloudflare_ips.sh'
#   action :nothing
# end

template '/etc/nginx/nginx.conf' do
  source 'nginx.conf.erb'
  mode '0644'
  owner 'root'
  group 'root'
  variables(
    user: node['magento']['nginx']['user'],
    w_processes: node['magento']['nginx']['worker_processes'],
    error_log: node['magento']['nginx']['error_log'],
    error_type: node['magento']['nginx']['error_log_type'],
    w_conns: node['magento']['nginx']['worker_connections'],
    sendfile: node['magento']['nginx']['sendfile'],
    l_subrequest: node['magento']['nginx']['log_subrequest'],
    tcp_nopush: node['magento']['nginx']['tcp_nopush'],
    tcp_nodelay: node['magento']['nginx']['tcp_nodelay'],
    etag: node['magento']['nginx']['etag'],
    kl_timeout: node['magento']['nginx']['keepalive_timeout'],
    types_hash_max_size: node['magento']['nginx']['types_hash_max_size'],
    resolver: node['magento']['nginx']['resolver'],
    gzip: node['magento']['nginx']['gzip'],
    server_tokens: node['magento']['nginx']['server_tokens'],
    fastcgi_read_timeout: node['magento']['nginx']['fastcgi_read_timeout'],
    cf_ips: node['magento']['nginx']['cloudflare_ips']
  )
end

service 'nginx' do
  action :nothing
  subscribes :restart, 'file[/etc/nginx/nginx.conf]', :immediately
  # notifies :notify, 'slack_notify[nginx_configured]', :immediately
end
#
directory '/root/.composer' do
  owner 'magento'
  group 'magento'
  mode '0755'
  action :create
end

cookbook_file '/root/.composer/auth.json' do
  source 'auth.json'
  owner 'magento'
  group 'magento'
  mode '0755'
  action :create
end
directory '/home/magento/public' do
  owner 'magento'
  group 'magento'
  mode '0755'
  action :create
end
# To get the Magento metapackage:
# https://magento.com/tech-resources/downloads/index/download/file_id/2876/category_id/2341/
execute 'install_magento' do
  cwd '/home/magento/public'
  command 'php -d max_execution_time=21600 /usr/local/bin/composer create-project --no-progress --repository-url=https://repo.magento.com/ magento/project-community-edition /home/magento/public'
  not_if { File.exist?('/home/magento/public/composer.json') }
end
# magento install  composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition <install-directory-name>

# You must set read-write permissions for the web server group
# cd /var/www/html/<magento install directory>
# %w[pub/static pub/media app/etc]
# find var generated vendor pub/static pub/media app/etc -type f -exec chmod g+w {} +
execute 'file_permitions' do
  cwd '/home/magento'
  command 'find var generated vendor pub/static pub/media app/etc -type f -exec chmod g+w {} +'
end
execute 'dir_permitions' do
  cwd '/home/magento'
  command 'find var generated vendor pub/static pub/media app/etc -type d -exec chmod g+ws {} +'
end
execute 'dir_permitions' do
  cwd '/home/magento'
  command 'chmod u+x bin/magento'
end

# find var generated vendor pub/static pub/media app/etc -type d -exec chmod g+ws {} +
# chown -R :www-data . # Ubuntu
# chmod u+x bin/magento

# installing Magento Command line
# bin/magento setup:install \
# --base-url=http://localhost/magento2ee \
# --db-host=localhost \
# --db-name=magento \
# --db-user=magento \
# --db-password=magento \
# --admin-firstname=admin \
# --admin-lastname=admin \
# --admin-email=admin@admin.com \
# --admin-user=admin \
# --admin-password=admin123 \
# --language=en_US \
# --currency=USD \
# --timezone=America/Chicago \
# --use-rewrites=1
# --amqp-host="<hostname>" --amqp-port="5672" --amqp-user="<user_name>" --amqp-password="<password>" --amqp-virtualhost="/"

# remote_file '/opt/install.sh' do
#   source 'http://software.virtualmin.com/gpl/scripts/install.sh'
#   owner 'root'
#   group 'root'
#   mode '0777'
#   action :create
# end
#
# execute 'install_virtuamin' do
#   command 'sh /opt/install.sh -f --bundle LEMP'
#   cwd '/opt'
#   action :run
#   # notifies :notify, 'slack_notify[virtualmin_installed]', :immediately
# end

# slack_notify 'virtualmin_installed' do
#   message "Virtualmin is Installed in Server #{host_name}"
#   webhook_url 'https://hooks.slack.com/services/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'
#   action :nothing
# end

# remote_file '/opt/csf.tgz' do
#   source 'https://download.configserver.com/csf.tgz'
#   owner 'root'
#   group 'root'
#   mode '0755'
#   action :create_if_missing
#   notifies :run, 'execute[unzip_csf]', :immediately
# end
#
# execute 'unzip_csf' do
#   command 'tar -xzf /opt/csf.tgz -C /opt/'
#   cwd '/opt'
#   action :run
#   notifies :run, 'execute[install_csf]', :immediately
# end
#
# execute 'install_csf' do
#   command 'sh /opt/csf/install.sh'
#   cwd '/opt/csf/'
#   action :run
# end
#
# template '/etc/csf/csf.conf' do
#   source 'csf.conf.erb'
#   mode '0600'
#   owner 'root'
#   group 'root'
#   variables(
#     tcp_in: node['magento']['csf']['TCP_IN'],
#     udp_in: node['magento']['csf']['UDP_IN'],
#     tcp6_in: node['magento']['csf']['TCP6_IN'],
#     udp6_in: node['magento']['csf']['UDP6_IN'],
#     restrict_logs: node['magento']['csf']['RESTRICT_SYSLOG'],
#     check_logs: node['magento']['csf']['SYSLOG_CHECK'],
#     date: node['magento']['date']
#   )
# end
#
# service 'csf' do
#   action %i(start enable)
#   notifies :run, 'execute[yum update -y]', :immediately
#   # notifies :notify, 'slack_notify[csf_installed]', :immediately
# end
#
# # slack_notify 'csf_installed' do
# #   message "Csf is Installed in Server #{host_name}"
# #   webhook_url 'https://hooks.slack.com/services/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'
# #   action :nothing
# # end
#
# # echo "Installing CSF webmin module"
# execute 'webmin_module' do
#   command '/usr/libexec/webmin/install-module.pl /usr/local/csf/csfwebmin.tgz'
#   action :run
# end
#

# #
# # slack_notify 'php_configured' do
# #   message "php is Configured in Server #{host_name}"
# #   webhook_url 'https://hooks.slack.com/services/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'
# #   action :nothing
# # end
#

# # slack_notify 'nginx_configured' do
# #   message "NGINX is Configured in Server #{host_name}"
# #   webhook_url 'https://hooks.slack.com/services/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'
# #   action :nothing
# # end
#
# %w[MariaDB-server MariaDB-client].each do |maria|
#   yum_package maria do
#     action :install
#   end
# end
#
# template '/etc/my.cnf' do
#   source 'my.cnf.erb'
#   mode '0644'
#   owner 'root'
#   group 'root'
#   variables(
#     datadir: node['magento']['mysql']['datadir'],
#     log_error: node['magento']['mysql']['log-error'],
#     s_lnk: node['magento']['mysql']['symbolic-links'],
#     max_conn: node['magento']['mysql']['max_connections'],
#     perf_schema: node['magento']['mysql']['performance_schema'],
#     sqry_log: node['magento']['mysql']['slow-query-log'],
#     sqry_log_file: node['magento']['mysql']['slow-query-log-file'],
#     long_query_time: node['magento']['mysql']['long_query_time'],
#     query_size: node['magento']['mysql']['query_cache_size'],
#     query_type: node['magento']['mysql']['query_cache_type'],
#     query_limit: node['magento']['mysql']['query_cache_limit'],
#     join_buffer_size: node['magento']['mysql']['join_buffer_size'],
#     thread_cache_size: node['magento']['mysql']['thread_cache_size'],
#     read_rnd_buffer_size: node['magento']['mysql']['read_rnd_buffer_size'],
#     read_buffer_size: node['magento']['mysql']['read_buffer_size'],
#     sort_buffer_size: node['magento']['mysql']['sort_buffer_size'],
#     key_buffer_size: node['magento']['mysql']['key_buffer_size'],
#     innodb_f_p_table: node['magento']['mysql']['innodb_file_per_table'],
#     innodb_p_size: node['magento']['mysql']['innodb_buffer_pool_size'],
#     innodb_p_instances: node['magento']['mysql']['innodb_buffer_pool_instances'],
#     innodb_ft_min_token_size: node['magento']['mysql']['innodb-ft-min-token-size'],
#     innodb_ft_enable_stopword: node['magento']['mysql']['innodb_ft_enable_stopword'],
#     ft_min_word_len: node['magento']['mysql']['ft_min_word_len'],
#     ft_stopword_file: node['magento']['mysql']['ft_stopword_file']
#     # innodb_log_file_size = 2G
#   )
# end
#
# service 'mariadb' do
#   action %i(start enable)
# end
#
# execute 'mkdir /opt/scripts' do
#   action :run
#   not_if { Dir.exist?('/opt/scripts') }
# end
# cookbook_file '/opt/scripts/cloudflare_ips.sh' do
#   source 'cloudflare_ips.sh'
#   owner 'root'
#   group 'root'
#   mode '0777'
#   action :create
# end
#
# execute 'echo "root: incidents@studio-one.am" >> /etc/aliases'
#
# execute 'newaliases'
#
# cron 'clamscan_report' do
#   action :create
#   minute '0'
#   hour '0'
#   day '*'
#   month '*'
#   weekday '*'
#   user 'root'
#   mailto 'incidents@studio-one.am'
#   command 'clamscan -i -r * / 2>&1 | mail -s "CRON CLAMSCAN" incidents@studio-one.am'
# end
#
# cron 'cloudflare_ips' do
#   action :create
#   minute '0'
#   hour '04'
#   day '*'
#   month '*'
#   weekday '*'
#   user 'root'
#   home '/opt/scripts'
#   mailto 'incidents@studio-one.am'
#   command 'sh /opt/scripts/cloudflare_ips.sh'
#   # notifies :notify, 'slack_notify[install_netdata]', :immediately
# end
# #
# # slack_notify 'install_netdata' do
# #   message "Installation of netdata is starting in the Server #{host_name}"
# #   webhook_url 'https://hooks.slack.com/services/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'
# #   action :nothing
# # end
#
# remote_file '/opt/kickstart.sh' do
#   source 'https://my-netdata.io/kickstart.sh'
#   owner 'root'
#   group 'root'
#   mode '0777'
#   action :create_if_missing
#   notifies :run, 'execute[install_netdata]', :immediately
#   # notifies :create, 'yum_repository[remi-remi]', :immediately
# end
#
# execute 'install_netdata' do
#   command 'sh /opt/kickstart.sh  --dont-wait --non-interactive'
#   cwd '/opt'
#   user 'root'
#   action :run
#   # notifies :notify, 'slack_notify[end_setup]', :immediately
# end
#
# # slack_notify 'end_setup' do
# #   message "The Setup is done, let's clean the Server #{host_name}"
# #   webhook_url 'https://hooks.slack.com/services/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'
# #   action :nothing
# # end
# # Deleting all remains to clean the system.
# file '/opt/csf.tgz' do
#   action :delete
# end
#
# file '/opt/install.sh' do
#   action :delete
# end
#
# file '/opt/composer.phar' do
#   action :delete
# end
#
# file '/opt/composer-setup.php' do
#   action :delete
# end
#
# directory '/opt/csf' do
#   recursive true
#   action :delete
#   notifies :run, 'execute[yum update -y]', :immediately
#   # notifies :notify, 'slack_notify[end_provision]', :immediately
# end
#
# slack_notify 'end_provision' do
#   message "The provision is done in Server #{host_name}"
#   webhook_url 'https://hooks.slack.com/services/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'
#   action :nothing
# end
